src/promql/src/extension_plan/instant_manipulator.rs

```diff
@@ -0,0 +1,721 @@
+    pub fn manipulate(&self, input: RecordBatch) -> ArrowResult<RecordBatch> {
+        let mut take_indices = Vec::with_capacity(input.num_rows());
+        // TODO(ruihang): maybe the input is not timestamp millisecond array
+        let ts_column = input
+            .column(self.time_index)
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap();
+
+        // prepare two vectors
+        let mut cursor = 0;
+        let aligned_ts = (self.start..=self.end)
+            .step_by(self.interval as usize)
+            .collect::<Vec<_>>();
+
+        // calculate the offsets to take
+        'next: for expected_ts in aligned_ts.iter().copied() {
+            // first, search toward end to see if there is matched timestamp
+            while cursor < ts_column.len() {
+                let curr = ts_column.value(cursor);
+                match curr.cmp(&expected_ts) {
+                    Ordering::Equal => {
+                        take_indices.push(Some(cursor as u64));
+                        continue 'next;
+                    }
+                    Ordering::Greater => break,
+                    Ordering::Less => {}
+                }
+                cursor += 1;
+            }
+            if cursor == ts_column.len() {
+                cursor -= 1;
+            }
+            // then, search backward to lookback
+            loop {
+                let curr = ts_column.value(cursor);
+                if curr + self.lookback_delta < expected_ts {
+                    // not found in lookback, leave this field blank.
+                    take_indices.push(None);
+                    break;
+                } else if curr < expected_ts && curr + self.lookback_delta >= expected_ts {
+                    // find the expected value, push and break
+                    take_indices.push(Some(cursor as u64));
+                    break;
+                } else if cursor == 0 {
+                    // reach the first value and not found in lookback, leave this field blank
+                    take_indices.push(None);
+                    break;
+                }
+                cursor -= 1;
+            }
+        }
+
+        // take record batch and replace the time index column
+        self.take_record_batch_optional(input, take_indices, aligned_ts)
+    }
+
+    /// Helper function to apply "take" on record batch.
+    fn take_record_batch_optional(
+        &self,
+        record_batch: RecordBatch,
+        take_indices: Vec<Option<u64>>,
+        aligned_ts: Vec<Millisecond>,
+    ) -> ArrowResult<RecordBatch> {
+        let indices_array = UInt64Array::from(take_indices);
+        let mut arrays = record_batch
+            .columns()
+            .iter()
+            .map(|array| compute::take(array, &indices_array, None))
+            .collect::<ArrowResult<Vec<_>>>()?;
+        arrays[self.time_index] = Arc::new(TimestampMillisecondArray::from(aligned_ts));
```

Comment:
Do we need to filter elements without value (whose index is None)? Does Prometheus keep timestamps that don't have any point in them?

Reply:
I think this comment is duplicated https://github.com/GreptimeTeam/greptimedb/pull/803#discussion_r1059297633